[AVR] Correctly set the pointer address space when constructing pointers to functions

dylanmckay opened the pull request:
NOTE: Pull request iterations:

* https://github.com/dylanmckay/rust/releases/tag/avr-use-correct-addrspace.0
* https://github.com/dylanmckay/rust/releases/tag/avr-use-correct-addrspace.1
* https://github.com/dylanmckay/rust/releases/tag/avr-use-correct-addrspace.2

This patch extends the existing `type_i8p` method so that it requires an
explicit address space to be specified. Before this patch, the
`type_i8p` method implcitily assumed the default address space, which is
not a safe transformation on all targets, namely AVR.

The Rust compiler already has support for tracking the "instruction
address space" on a per-target basis. This patch extends the code
generation routines so that an address space must always be specified.

In my estimation, around 15% of the callers of `type_i8p` produced
invalid code on AVR due to the loss of address space prior to LLVM final
code generation. This would lead to unavoidable assertion errors
relating to invalid bitcasts.

With this patch, the address space is always either 1) explicitly preserved
from the input type, or 2) explicitly set to the instruction address
space because the logic is dealing with functions which must be placed
there, or 3) explicitly set to the default address space 0 because the
logic can only operate on data space pointers and thus we keep the
existing semantics of assuming the default, "data" address space.

dylanmckay committed:
[AVR] Correctly set the pointer address space when constructing pointers to functions

This patch extends the existing `type_i8p` method so that it requires an
explicit address space to be specified. Before this patch, the
`type_i8p` method implcitily assumed the default address space, which is
not a safe transformation on all targets, namely AVR.

The Rust compiler already has support for tracking the "instruction
address space" on a per-target basis. This patch extends the code
generation routines so that an address space must always be specified.

In my estimation, around 15% of the callers of `type_i8p` produced
invalid code on AVR due to the loss of address space prior to LLVM final
code generation. This would lead to unavoidable assertion errors
relating to invalid bitcasts.

With this patch, the address space is always either 1) explicitly set to
the instruction address space because the logic is dealing with functions
which must be placed there, or 2) explicitly set to the default address
space 0 because the logic can only operate on data space pointers and thus
we keep the existing semantics of assuming the default, "data" address space.

rust-highfive commented:
r? @matthewjasper

(rust_highfive has picked a reviewer for you, use r? to override)

dylanmckay commented on src/librustc_codegen_llvm/builder.rs:
```diff
@@ -1248,7 +1248,8 @@ impl Builder<'a, 'll, 'tcx> {
 
-        let ptr = self.pointercast(ptr, self.cx.type_i8p());
+        let ptr = self
+            .pointercast(ptr, self.cx.type_i8p(self.cx.data_layout().instruction_address_space));
```
TODO: this can be turned into a "preserving" address space from `ptr` rather than throwing it away because it _should_ be the same as `self.cx.data_layout().instruction_address_space)`

jfrimmel commented on src/librustc_target/abi/mod.rs:
```diff
@@ -744,6 +749,18 @@ impl FieldsShape {
+/// An identifier that specifies the address space that some operation
+/// should operate on. Special address spaces have an effect on code generation,
+/// depending on the target and the address spaces it implements.
+#[derive(Copy, Clone, Debug, PartialEq, Eq, PartialOrd, Ord)]
+pub struct AddressSpace(pub u32);
+
+impl Default for AddressSpace {
+    fn default() -> Self {
+        AddressSpace(0)
+    }
+}
```
One could derive the `Default` impl: 
```suggestion
#[derive(Copy, Clone, Debug, Default, PartialEq, Eq, PartialOrd, Ord)]
pub struct AddressSpace(pub u32);
```

dylanmckay replied:
TIL the standard primitives also implement `Default`, will change.

dylanmckay replied:
Fixed in da620823f5fabba1b891217820f7b5db7f2225c2.

rust-highfive commented:
The job `mingw-check` of your PR [failed](https://dev.azure.com/rust-lang/rust/_build/results?buildId=31800&view=logs&jobId=74054d28-c774-5fab-ecc2-352fe71b230e) ([pretty log](https://rust-lang.github.io/rust-log-analyzer/log-viewer/#https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_apis/build/builds/31800/logs/67), [raw log](https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_apis/build/builds/31800/logs/67)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
##[section]Starting: Linux mingw-check
##[section]Starting: Initialize job
Agent name: 'Azure Pipelines 23'
Agent machine name: 'fv-az578'
Current agent version: '2.170.1'
##[group]Operating System
16.04.6
LTS
LTS
##[endgroup]
##[group]Virtual Environment
Environment: ubuntu-16.04
Version: 20200604.1
Included Software: https://github.com/actions/virtual-environments/blob/ubuntu16/20200604.1/images/linux/Ubuntu1604-README.md
##[endgroup]
Agent running as: 'vsts'
Prepare build directory.
Set build variables.
Download all required tasks.
Download all required tasks.
Downloading task: Bash (3.163.3)
Checking job knob settings.
   Knob: AgentToolsDirectory = /opt/hostedtoolcache Source: ${AGENT_TOOLSDIRECTORY} 
   Knob: AgentPerflog = /home/vsts/perflog Source: ${VSTS_AGENT_PERFLOG} 
Start tracking orphan processes.
##[section]Finishing: Initialize job
##[section]Starting: Configure Job Name
==============================================================================
---
========================== Starting Command Output ===========================
[command]/bin/bash --noprofile --norc /home/vsts/work/_temp/e93121e5-e57a-4523-8b4d-fe71a9dfb9ac.sh

##[section]Finishing: Disable git automatic line ending conversion
##[section]Starting: Checkout rust-lang/rust@refs/pull/73270/merge to s
Task         : Get sources
Description  : Get sources from a repository. Supports Git, TfsVC, and SVN repositories.
Version      : 1.0.0
Author       : Microsoft
---
##[command]git remote add origin https://github.com/rust-lang/rust
##[command]git config gc.auto 0
##[command]git config --get-all http.https://github.com/rust-lang/rust.extraheader
##[command]git config --get-all http.proxy
##[command]git -c http.extraheader="AUTHORIZATION: basic ***" fetch --force --tags --prune --progress --no-recurse-submodules --depth=2 origin +refs/heads/*:refs/remotes/origin/* +refs/pull/73270/merge:refs/remotes/pull/73270/merge
---
 ---> f883e675ad62
Step 6/7 : ENV RUN_CHECK_WITH_PARALLEL_QUERIES 1
 ---> Using cache
 ---> c0b156eb069c
Step 7/7 : ENV SCRIPT python3 ../x.py test src/tools/expand-yaml-anchors &&            python3 ../x.py check --target=i686-pc-windows-gnu --host=i686-pc-windows-gnu &&            python3 ../x.py build --stage 0 src/tools/build-manifest &&            python3 ../x.py test --stage 0 src/tools/compiletest &&            python3 ../x.py test src/tools/tidy &&            python3 ../x.py doc --stage 0 src/libstd &&            /scripts/validate-toolstate.sh
 ---> 8541bab6b38c
Successfully built 8541bab6b38c
Successfully tagged rust-ci:latest
Built container sha256:8541bab6b38c07f1b7eb787539b9cbe93daa6ac4458d3d7bd8a8921622a14ba1
---
    Checking rustc_span v0.0.0 (/checkout/src/librustc_span)
    Checking rustc_errors v0.0.0 (/checkout/src/librustc_errors)
    Checking rustc_feature v0.0.0 (/checkout/src/librustc_feature)
    Checking rustc_parse_format v0.0.0 (/checkout/src/librustc_parse_format)
    Checking chalk-rust-ir v0.10.0
    Checking rustc_query_system v0.0.0 (/checkout/src/librustc_query_system)
    Checking rustc_hir v0.0.0 (/checkout/src/librustc_hir)
    Checking rustc_session v0.0.0 (/checkout/src/librustc_session)
    Checking chalk-solve v0.10.0
    Checking rustc_parse v0.0.0 (/checkout/src/librustc_parse)
---
configure: dist.missing-tools   := True
configure: build.submodules     := False
configure: llvm.assertions      := True
configure: build.locked-deps    := True
configure: build.configure-args := ['--enable-sccache', '--disable-manage-submodu ...
configure: writing `config.toml` in current directory
configure: 
configure: run `python /checkout/x.py --help`
configure: 
---
Hugepagesize:       2048 kB
DirectMap4k:      143296 kB
DirectMap2M:     4050944 kB
DirectMap1G:     5242880 kB
+ python3 ../x.py test src/tools/expand-yaml-anchors
Ensuring the YAML anchors in the GitHub Actions config were expanded
Ensuring the YAML anchors in the GitHub Actions config were expanded
Building stage0 tool expand-yaml-anchors (x86_64-unknown-linux-gnu)
   Compiling unicode-xid v0.2.0
   Compiling syn v1.0.11
   Compiling linked-hash-map v0.5.2
   Compiling lazy_static v1.4.0
   Compiling lazy_static v1.4.0
   Compiling yaml-rust v0.4.3
   Compiling quote v1.0.2
   Compiling thiserror-impl v1.0.5
   Compiling thiserror v1.0.5
   Compiling yaml-merge-keys v0.4.0
   Compiling expand-yaml-anchors v0.1.0 (/checkout/src/tools/expand-yaml-anchors)
Build completed successfully in 0:00:30
+ python3 ../x.py check --target=i686-pc-windows-gnu --host=i686-pc-windows-gnu
    Finished dev [unoptimized] target(s) in 0.18s
Checking rustdoc artifacts (x86_64-unknown-linux-gnu -> i686-pc-windows-gnu)
---
    Checking rustc_parse_format v0.0.0 (/checkout/src/librustc_parse_format)
    Checking chalk-rust-ir v0.10.0
    Checking rustc_ast_pretty v0.0.0 (/checkout/src/librustc_ast_pretty)
    Checking rustc_hir v0.0.0 (/checkout/src/librustc_hir)
    Checking rustc_query_system v0.0.0 (/checkout/src/librustc_query_system)
    Checking chalk-solve v0.10.0
    Checking rustc_parse v0.0.0 (/checkout/src/librustc_parse)
    Checking rustc_hir_pretty v0.0.0 (/checkout/src/librustc_hir_pretty)
    Checking rustc_ast_lowering v0.0.0 (/checkout/src/librustc_ast_lowering)
---
skip untracked path src/doc/rust-by-example/ during rustfmt invocations
skip untracked path src/llvm-project/ during rustfmt invocations
Diff in /checkout/src/librustc_codegen_llvm/builder.rs at line 1248:
 
         let lifetime_intrinsic = self.cx.get_intrinsic(intrinsic);
-        let ptr = self
-        let ptr = self
-            .pointercast(ptr, self.cx.type_i8p(self.cx.address_space_of_value(ptr)));
+        let ptr = self.pointercast(ptr, self.cx.type_i8p(self.cx.address_space_of_value(ptr)));
         self.call(lifetime_intrinsic, &[self.cx.const_u64(size), ptr], None);
 
 
Running `"/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/rustfmt" "--config-path" "/checkout" "--edition" "2018" "--unstable-features" "--skip-children" "--check" "/checkout/src/librustc_codegen_llvm/builder.rs"` failed.
If you're running `tidy`, try again with `--bless`. Or, if you just want to format code, run `./x.py fmt` instead.
Build completed unsuccessfully in 0:00:39
== clock drift check ==
  local time: Fri Jun 12 12:11:54 UTC 2020
  network time: Fri, 12 Jun 2020 12:11:54 GMT
  network time: Fri, 12 Jun 2020 12:11:54 GMT
== end clock drift check ==

##[error]Bash exited with code '1'.
##[section]Finishing: Run build
##[section]Starting: Checkout rust-lang/rust@refs/pull/73270/merge to s
Task         : Get sources
Description  : Get sources from a repository. Supports Git, TfsVC, and SVN repositories.
Version      : 1.0.0
Author       : Microsoft
Author       : Microsoft
Help         : [More Information](https://go.microsoft.com/fwlink/?LinkId=798199)
==============================================================================
Cleaning any cached credential from repository: rust-lang/rust (GitHub)
##[section]Finishing: Checkout rust-lang/rust@refs/pull/73270/merge to s
Cleaning up task key
Start cleaning up orphan processes.
Terminate orphan process: pid (3661) (python)
##[section]Finishing: Finalize Job
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@rust-lang/infra`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

shepmaster commented on src/librustc_target/abi/mod.rs:
```diff
@@ -744,6 +749,12 @@ impl FieldsShape {
+/// An identifier that specifies the address space that some operation
+/// should operate on. Special address spaces have an effect on code generation,
+/// depending on the target and the address spaces it implements.
+#[derive(Copy, Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Default)]
+pub struct AddressSpace(pub u32);
```
It may be better to avoid `Default` and instead call that method `data`. It'd be shorter at the call sites and potentially more expressive.

shepmaster replied:
Assuming that it _is_ the "data" address space...

dylanmckay replied:
Done in 326b9d8fffd726469577ad0f59b25dadbfe068a9

dylanmckay committed:
[AVR] Ensure that function pointers stored within aggregates are annotated with the correct space

Before this patch, a function pointer stored within an aggregate like a
struct or an enum would always have the default address space `0`.

This patch removes this assumption and instead, introspects the inner
type being pointed at, storing the target address space in the PointeeInfo
struct so that downstream users may query it.

rust-highfive commented:
The job `mingw-check` of your PR [failed](https://dev.azure.com/rust-lang/rust/_build/results?buildId=32337&view=logs&jobId=74054d28-c774-5fab-ecc2-352fe71b230e) ([pretty log](https://rust-lang.github.io/rust-log-analyzer/log-viewer/#https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_apis/build/builds/32337/logs/65), [raw log](https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_apis/build/builds/32337/logs/65)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
##[section]Starting: Linux mingw-check
##[section]Starting: Initialize job
Agent name: 'Azure Pipelines 2'
Agent machine name: 'fv-az578'
Current agent version: '2.170.1'
##[group]Operating System
16.04.6
LTS
LTS
##[endgroup]
##[group]Virtual Environment
Environment: ubuntu-16.04
Version: 20200614.1
Included Software: https://github.com/actions/virtual-environments/blob/ubuntu16/20200614.1/images/linux/Ubuntu1604-README.md
##[endgroup]
Agent running as: 'vsts'
Prepare build directory.
Set build variables.
Download all required tasks.
Download all required tasks.
Downloading task: Bash (3.163.3)
Checking job knob settings.
   Knob: AgentToolsDirectory = /opt/hostedtoolcache Source: ${AGENT_TOOLSDIRECTORY} 
   Knob: AgentPerflog = /home/vsts/perflog Source: ${VSTS_AGENT_PERFLOG} 
Start tracking orphan processes.
##[section]Finishing: Initialize job
##[section]Starting: Configure Job Name
==============================================================================
---
========================== Starting Command Output ===========================
[command]/bin/bash --noprofile --norc /home/vsts/work/_temp/00ef6593-11a3-48f9-889d-526210031fca.sh

##[section]Finishing: Disable git automatic line ending conversion
##[section]Starting: Checkout rust-lang/rust@refs/pull/73270/merge to s
Task         : Get sources
Description  : Get sources from a repository. Supports Git, TfsVC, and SVN repositories.
Version      : 1.0.0
Author       : Microsoft
---
##[command]git remote add origin https://github.com/rust-lang/rust
##[command]git config gc.auto 0
##[command]git config --get-all http.https://github.com/rust-lang/rust.extraheader
##[command]git config --get-all http.proxy
##[command]git -c http.extraheader="AUTHORIZATION: basic ***" fetch --force --tags --prune --progress --no-recurse-submodules --depth=2 origin +refs/heads/*:refs/remotes/origin/* +refs/pull/73270/merge:refs/remotes/pull/73270/merge
---
 ---> a9ec21d337b3
Step 6/7 : ENV RUN_CHECK_WITH_PARALLEL_QUERIES 1
 ---> Using cache
 ---> 5ff2c13d8dba
Step 7/7 : ENV SCRIPT python3 ../x.py test src/tools/expand-yaml-anchors &&            python3 ../x.py check --target=i686-pc-windows-gnu --host=i686-pc-windows-gnu &&            python3 ../x.py build --stage 0 src/tools/build-manifest &&            python3 ../x.py test --stage 0 src/tools/compiletest &&            python3 ../x.py test src/tools/tidy &&            python3 ../x.py doc --stage 0 src/libstd &&            /scripts/validate-toolstate.sh
 ---> 6b931e755c7e
Successfully built 6b931e755c7e
Successfully tagged rust-ci:latest
Built container sha256:6b931e755c7ea1f69816640eb9df74fafd40a545d0e5aa8341d35009dabb0f3c
---
    Checking rustc_span v0.0.0 (/checkout/src/librustc_span)
    Checking rustc_errors v0.0.0 (/checkout/src/librustc_errors)
    Checking rustc_feature v0.0.0 (/checkout/src/librustc_feature)
    Checking rustc_parse_format v0.0.0 (/checkout/src/librustc_parse_format)
    Checking chalk-rust-ir v0.10.0
    Checking rustc_query_system v0.0.0 (/checkout/src/librustc_query_system)
    Checking rustc_hir v0.0.0 (/checkout/src/librustc_hir)
    Checking rustc_session v0.0.0 (/checkout/src/librustc_session)
    Checking chalk-solve v0.10.0
    Checking rustc_parse v0.0.0 (/checkout/src/librustc_parse)
---
configure: build.cargo-native-static := True
configure: rust.channel         := nightly
configure: rust.dist-src        := False
configure: build.locked-deps    := True
configure: build.configure-args := ['--enable-sccache', '--disable-manage-submodu ...
configure: writing `config.toml` in current directory
configure: 
configure: run `python /checkout/x.py --help`
configure: 
---
Hugepagesize:       2048 kB
DirectMap4k:      112576 kB
DirectMap2M:     4081664 kB
DirectMap1G:     5242880 kB
+ python3 ../x.py test src/tools/expand-yaml-anchors
Ensuring the YAML anchors in the GitHub Actions config were expanded
Ensuring the YAML anchors in the GitHub Actions config were expanded
Building stage0 tool expand-yaml-anchors (x86_64-unknown-linux-gnu)
   Compiling unicode-xid v0.2.0
   Compiling syn v1.0.11
   Compiling linked-hash-map v0.5.2
   Compiling lazy_static v1.4.0
   Compiling lazy_static v1.4.0
   Compiling yaml-rust v0.4.3
   Compiling quote v1.0.2
   Compiling thiserror-impl v1.0.5
   Compiling thiserror v1.0.5
   Compiling yaml-merge-keys v0.4.0
   Compiling expand-yaml-anchors v0.1.0 (/checkout/src/tools/expand-yaml-anchors)
Build completed successfully in 0:00:28
+ python3 ../x.py check --target=i686-pc-windows-gnu --host=i686-pc-windows-gnu
    Finished dev [unoptimized] target(s) in 0.17s
Checking rustdoc artifacts (x86_64-unknown-linux-gnu -> i686-pc-windows-gnu)
---
    Checking rustc_parse_format v0.0.0 (/checkout/src/librustc_parse_format)
    Checking chalk-rust-ir v0.10.0
    Checking rustc_ast_pretty v0.0.0 (/checkout/src/librustc_ast_pretty)
    Checking rustc_hir v0.0.0 (/checkout/src/librustc_hir)
    Checking rustc_query_system v0.0.0 (/checkout/src/librustc_query_system)
    Checking chalk-solve v0.10.0
    Checking rustc_parse v0.0.0 (/checkout/src/librustc_parse)
    Checking rustc_hir_pretty v0.0.0 (/checkout/src/librustc_hir_pretty)
    Checking rustc_ast_lowering v0.0.0 (/checkout/src/librustc_ast_lowering)
---
skip untracked path cpu-usage.csv during rustfmt invocations
skip untracked path src/doc/book/ during rustfmt invocations
skip untracked path src/doc/rust-by-example/ during rustfmt invocations
skip untracked path src/llvm-project/ during rustfmt invocations
Diff in /checkout/src/librustc_middle/ty/layout.rs at line 2282:
         };
 
 
-        debug!("pointee_info_at (offset={:?}, type kind: {:?}) => {:?}", offset, this.ty.kind, pointee_info);
+        debug!(
+            "pointee_info_at (offset={:?}, type kind: {:?}) => {:?}",
+            offset, this.ty.kind, pointee_info
 
         pointee_info
     }
     }
Running `"/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/rustfmt" "--config-path" "/checkout" "--edition" "2018" "--unstable-features" "--skip-children" "--check" "/checkout/src/librustc_middle/ty/layout.rs"` failed.
If you're running `tidy`, try again with `--bless`. Or, if you just want to format code, run `./x.py fmt` instead.
Build completed unsuccessfully in 0:00:42
== clock drift check ==
  local time: Fri Jun 19 08:33:40 UTC 2020
  network time: Fri, 19 Jun 2020 08:33:40 GMT
  network time: Fri, 19 Jun 2020 08:33:40 GMT
== end clock drift check ==

##[error]Bash exited with code '1'.
##[section]Finishing: Run build
##[section]Starting: Checkout rust-lang/rust@refs/pull/73270/merge to s
Task         : Get sources
Description  : Get sources from a repository. Supports Git, TfsVC, and SVN repositories.
Version      : 1.0.0
Author       : Microsoft
Author       : Microsoft
Help         : [More Information](https://go.microsoft.com/fwlink/?LinkId=798199)
==============================================================================
Cleaning any cached credential from repository: rust-lang/rust (GitHub)
##[section]Finishing: Checkout rust-lang/rust@refs/pull/73270/merge to s
Cleaning up task key
Start cleaning up orphan processes.
Terminate orphan process: pid (4300) (python)
##[section]Finishing: Finalize Job
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@rust-lang/infra`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

bors commented:
:umbrella: The latest upstream changes (presumably #73504) made this pull request unmergeable. Please resolve the merge conflicts.

bors commented:
:umbrella: The latest upstream changes (presumably #73528) made this pull request unmergeable. Please resolve the merge conflicts.

shepmaster commented:
Can we remove the hacks we made to the formatters with this PR?

dylanmckay commented:
> @shepmaster: Can we remove the hacks we made to the formatters with this PR?

I reverted 91bff8ccdfddd119a2239e625bfb52703844f227 from Rust master locally on top of this patch but then it breaks libcore

```
LLVM ERROR: Cannot select: t9: i16 = addrspacecast[1 -> 0] t8
  t8: i16,ch = load<(load 2 from %ir.5, align 1, !alias.scope !265)> t0, t7, undef:i16
    t7: i16,ch = load<(dereferenceable load 2 from %ir.4, align 1)> t0, t2, undef:i16
      t2: i16,ch = CopyFromReg t0, Register:i16 %19
        t1: i16 = Register %19
      t6: i16 = undef
    t6: i16 = undef
In function: _ZN42_$LT$$RF$T$u20$as$u20$core..fmt..Debug$GT$3fmt17h2dc31f322889dd28E
```

So it looks like we cannot remove the hacks we made to the formatters in 91bff8ccdfddd119a2239e625bfb52703844f227 with this PR.

nagisa commented:
r? @nagisa I’ll review this tomorrow.

bors commented:
:umbrella: The latest upstream changes (presumably #73669) made this pull request unmergeable. Please resolve the merge conflicts.

nagisa reviewed (commented):
This changeset is lacking a test (of any kind, really, but a codegen test for something like function ptr to *mut () back to function ptr would be good). That said, I wonder if it is even possible to write one? The GPU targets could be a good way to get rust/llvm generate code for a split-address space architecture while AVR target is not yet mainline.

Otherwise most of the changes here look pretty mechanical and difficult to get wrong. I left some inline comments with suggestions on how this could be improvement, but all of them are fairly minor. Let me know if you don’t intend to address them here.

shepmaster commented:
> while AVR target is not yet mainline

Can you expand a bit on what more “mainline” would mean? The current nightly has an AVR target.

nagisa commented:
> Can you expand a bit on what more “mainline” would mean? The current nightly has an AVR target.

Oh, I must have checked for a target in an older checkout/version. In that case please do by all means write a test :)

eddyb commented:
Why is any of this necessary, does LLVM not allow the equivalent of casting a `fn()` to `usize` and then to `fn()` again (in order to call it), through pointer types? Seems like a design flaw to me (assuming both code and data pointers are the same size).

Also, the PR description talks about `i8p` but that's just a special-case of `type_ptr_to` (i.e. creating a LLVM pointer type), seems to me like the latter is the actual change here. Or is the point that "cast to `i8*`" needs to be address-space-polymorphic?

Most things that take `i8*` take them in some API that uses address space `0`, so I don't think "`i8*` with non-`0` address space" makes sense (or would work, except for us aggressively bitcasting on load/store/call due to HRTB subtyping issues, *sigh*).

Have you tried just replacing `pointercast` with bitcast`, in a few places, or everywhere? Probably the easiest thing to try would be to change `fn pointercast` in the builder to just call `bitcast`, so that you don't have to change any of the uses.

shepmaster commented:
> Have you tried just replacing `pointercast` with `bitcast`, in a few places, or everywhere? Probably the easiest thing to try would be to change `fn pointercast` in the builder to just call `bitcast`, so that you don't have to change any of the uses.

Doing so yields:

```
Invalid bitcast
  %57 = bitcast i8* %54 to i1 (%"fmt::::Opaque"*, %"fmt::Formatter"*) addrspace(1)*, !dbg !2679
in function _ZN4core3fmt3run17h34a80d70156b4c0eE
```

dylanmckay commented:
I tried it too and got the same result as @shepmaster. I ended up using a debugging to find all of the required patches and then halving the size of the patch, whilst removing all of the address space "preservation" logic as it could be trivially replaced.

I could reduce the size of the patch further by removing the newly added parameter from `type_i8p` that requires all callers to explicitly specify an address space. This could be moved to a `type_i8p_ext` like method, which would get rid of a lot of the diff in all but the few places that require passing the instruction address space.

I will see if I can reduce the size of the patch further by replacing `pointercast` instructions with `bitcast` instructions.

dylanmckay commented on src/rustllvm/RustWrapper.cpp:
```diff
@@ -1007,6 +1008,28 @@ extern "C" int64_t LLVMRustDIBuilderCreateOpPlusUconst() {
   return dwarf::DW_OP_plus_uconst;
 }
 
+// Returns `-1` if the value is not a pointer, otherwise returns the actual address space
```
I've since deleted this function and its doc comment, explicitly setting the address space rather than preserving it via existing types.

dylanmckay commented on src/librustc_middle/ty/layout.rs:
```diff
@@ -2148,16 +2148,32 @@ where
+                })
+            }
+            ty::FnPtr(fn_sig) if offset.bytes() == 0 => {
+                cx.layout_of(cx.tcx().mk_fn_ptr(fn_sig)).to_result().ok().map(|layout| {
```
Good suggestion, I have applied and amended into the original comment.

dylanmckay commented on src/librustc_middle/ty/layout.rs:
```diff
@@ -2148,16 +2148,32 @@ where
-
             ty::Ref(_, ty, mt) if offset.bytes() == 0 => {
+                let address_space = addr_space_of_ty(ty);
+
```
Fixed and amended into the original commit

dylanmckay commented on src/librustc_codegen_ssa/mir/place.rs:
```diff
@@ -388,7 +398,10 @@ impl<'a, 'tcx, V: CodegenObject> PlaceRef<'tcx, V> {
-        downcast.llval = bx.pointercast(downcast.llval, bx.cx().type_ptr_to(variant_ty));
+        downcast.llval = bx.pointercast(
+            downcast.llval,
+            bx.cx().type_ptr_to(variant_ty, bx.cx().address_space_of_value(downcast.llval)),
```
I was thinking of a good way to abstract it out, this is a good suggestion.

Although I've since removed the preservation aspect of this pull request, opting to pass the address space explicitly, and only in the few places it it actually required.

dylanmckay commented:
NOTE: I'm keeping tags of all of the iterations of this patch so we can compare and/or pull anything back in.

nagisa commented:
@dylanmckay is this still a draft? The implementation looks good to me…

dylanmckay commented:
@nagisa I haven't got around to writing it up properly, but yeah the implementation is no longer a draft - this is about as minimal as I could get the patch using the "stop using pointercast, always use bitcast" approach. I prefer this current approach to the boilerplate-heavy one located at the [`avr-use-correct-addrspace.0`](https://github.com/dylanmckay/rust/releases/tag/avr-use-correct-addrspace.0) tag.

The only thing left before sending this PR back to review is... writing tests for it. I will flip this out of draft once that is done.

dylanmckay commented:
I've added a few tests. Whilst writing the tests, I noticed that I no longer needed to apply @eddyb's [suggestion](https://github.com/rust-lang/rust/pull/73270#issuecomment-649896398) that changed all `pointercast`s to `bitcast`s. It seems like the fixes made in the most recent commit made the `pointercast` modification redundant - I can compile core without the fix in dylanmckay/rust@a5500b0aef5271548221100176af4bd385ab7ac4. I have thusly dropped that commit from the MR. If the fix is desired later it can be found on tag [`avr-use-correct-addrspace.3`](https://github.com/dylanmckay/rust/releases/tag/avr-use-correct-addrspace.3).

Back to review.

nagisa commented:
@bors r+

bors commented:
:pushpin: Commit 5581ce6c10ae0b4e6503db0081e2defd7ef829ff has been approved by `nagisa`

<!-- @bors r=nagisa 5581ce6c10ae0b4e6503db0081e2defd7ef829ff -->
<!-- homu: {"type":"Approved","sha":"5581ce6c10ae0b4e6503db0081e2defd7ef829ff","approver":"nagisa"} -->

bors commented:
:hourglass: Testing commit 5581ce6c10ae0b4e6503db0081e2defd7ef829ff with merge e22b61bff0bdd08be7665607cb7be3748c8a35d2...
<!-- homu: {"type":"BuildStarted","head_sha":"5581ce6c10ae0b4e6503db0081e2defd7ef829ff","merge_sha":"e22b61bff0bdd08be7665607cb7be3748c8a35d2"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/897897609), [checks-azure](https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=35452)
Approved by: nagisa
Pushing e22b61bff0bdd08be7665607cb7be3748c8a35d2 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"nagisa","base_ref":"master","builders":{"checks-azure":"https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=35452","checks-actions":"https://github.com/rust-lang-ci/rust/runs/897897609"},"merge_sha":"e22b61bff0bdd08be7665607cb7be3748c8a35d2"} -->

RalfJung commented on src/librustc_middle/ty/layout.rs:
```diff
@@ -2166,16 +2166,31 @@ where
+            ty::FnPtr(fn_sig) if offset.bytes() == 0 => {
+                cx.layout_of(cx.tcx().mk_fn_ptr(fn_sig)).to_result().ok().map(|layout| {
+                    PointeeInfo {
+                        size: layout.size,
+                        align: layout.align.abi,
```
FWIW this makes very little sense. This sets the *pointee* size and alignment of function pointers to be the *pointer* size and alignment, i.e. on x86-64 this will be size 8 alignment 8, which is just completely wrong -- function pointers do not point to 8-aligned data.

Luckily these two fields are, from what I can tell, completely unused if `safe` is `None` (which does make me wonder why they are not inside the same `Option`).

RalfJung replied:
Actually sounds like the entire approach of using `PointeeInfo` to carry address space information is fundamentally broken; see [this Zulip discussion](https://rust-lang.zulipchat.com/#narrow/stream/131828-t-compiler/topic/Fn.20Ptr.20pointee.20info.20is.20nonsense.3F).

RalfJung commented on src/test/codegen/avr/avr-func-addrspace.rs:
```diff
@@ -0,0 +1,93 @@
+    unsafe {
+        STORAGE_FOO(&1, &mut buf);
+    }
+}
```
What happens when you store a function pointer in a type like `Result<&'static u8, fn()>`? (Example due to @eddyb)

In LLVM this type gets translated as `{ i64, ptr }`, where the first field indicates Ok vs Err and the 2nd field contains the data. So it is fundamentally impossible to have a dedicated function pointer address space here. Doesn't sound like the approach taken in this PR can possibly work here?
